urlf91495fd27_Rembau_complie_tgz-pJ8-matcher_MatchJ8/src/regexpToNFA/MainClass.java

```diff
@@ -0,0 +1,258 @@
+package regexpToNFA;
+
+import java.util.Enumeration;
+import java.util.Hashtable;
+import java.util.LinkedList;
+import java.util.TreeSet;
+
+import NFAToDNA.*;
+
+/**
+ * @author 木木（茂）
+ * 此类实现了从正则表达式到NFA的转换
+ * 日期：2012.4.14
+ */		
+public class MainClass {
+	private String regexp;
+	private int note=1;
+	private Unit firstUnit;
+	private LinkedList<Unit> unitList = new LinkedList<Unit>();//所有单元
+	private LinkedList<Transfer> transferList = new LinkedList<Transfer>();  //
+	private Hashtable<Integer,Node> nodeList = new Hashtable<Integer,Node>();
+	private TreeSet<String> inputChar = new TreeSet<String>();
+	private LinkedList<Integer> handledUnit =new LinkedList<Integer>(); //已处理的单元的开始int
+	public MainClass(String re){
+		this.regexp=re;
+		this.regexp=initRegexp();
+		System.out.println(this.regexp);
+		handle();
+		createNFA();
+	}
+	public Hashtable<Integer,Node> getNodeList(){
+		return this.nodeList;
+	}
+	public TreeSet<String> getInputChar(){
+		return this.inputChar;
+	} 
+	/**
+	 * 初始化正则表达式，添加括号，区分单元  如(a|b)\**ab ->(((a)|(b))(\*)*(a)(b))
+	 * 特殊部分有形如\w ，形如[a-b]
+	 */
+	public String initRegexp(){
+		char[] words = regexp.toCharArray();
+		StringBuilder sb=new StringBuilder();
+		for(int i=0;i<words.length;i++){
+			if(Syntax.isKeyWord(words[i])){
+				sb.append(words[i]);
+			}  else if(Syntax.isThransMean(words[i])){
+				if(i<words.length-1){
+					sb.append('(');
+					sb.append(words[i]);
+					sb.append(words[i+1]);
+					sb.append(')');
+					i++;
+				} else {
+					System.out.println("错误");
+					System.exit(1);
+				}
+			} else if(Syntax.isIncludeLeft(words[i])){
+				sb.append('(');
+				boolean mark=true;
+				do {
+					if(i>=words.length){
+						System.out.println("错误");
+						System.exit(1);
+					}
+					sb.append(words[i]);
+					if(Syntax.isIncludeRight(words[i])){
+						mark=false;
+					}
+					i++;
+				} while(mark);
+				i--;
+				sb.append(')');
+			} else{
+				sb.append('(');
+				sb.append(words[i]);
+				sb.append(')');
+			}
+		}
+		return new String(sb);
+	}
+	int i=0;
+	public void handle(){
+		handle(null,null,0,0);
+	}
+	/**
+	 * 构建所有的Unit
+	 * @param father：当前单元的父单元
+	 * @param brother：当前单元的同一级的前一个单元如：（（a）（b））a就是b的brother
+	 * @param markNotNextUnit：标记是否应该执行brother.setNextUnit(unit);此操作只有同级的元素之间存在.
+	 * 每一级unit_now初始化为null,既第一个unit的brother为null,当遇到‘|’时，设置markNotNextUnit为1，第一
+	 * 个unit过后恢复这个值，即‘|’符号后面的第一个unit不应有brother
+	 * @param markAddInnerFistUnit：标记是否应该执行 father.addInnerFirstUnit(unit);unit有子unit时
+	 * 子unit中的第一个unit应该执行这个操作，又或当遇到‘|’时，此符号的后面一个unit也应执行该操作
+	 * @return
+	 */
+	public Unit handle(Unit father,Unit brother,int markNotNextUnit,int markAddInnerFistUnit){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.